crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,956 @@
+    ///
+    /// If you only care to know about changes once all of them have applied
+    /// (including the total) listen to a clone of this signal.
+    #[builder(private)]
+    pub rooms_updated_broadcaster:
+        futures_signals::signal::Broadcaster<futures_signals::signal::Receiver<()>>,
+}
+
+/// the default name for the full sync view
+pub const FULL_SYNC_VIEW_NAME: &str = "full-sync";
+
+impl SlidingSyncViewBuilder {
+    /// Create a Builder set up for full sync
+    pub fn default_with_fullsync() -> Self {
+        Self::default().name(FULL_SYNC_VIEW_NAME).sync_mode(SlidingSyncMode::FullSync).to_owned()
+    }
+
+    /// Build the view
+    pub fn build(&mut self) -> Result<SlidingSyncView, SlidingSyncViewBuilderError> {
+        let (sender, receiver) = futures_signals::signal::channel(());
+        self.rooms_updated_signal = Some(sender);
+        self.rooms_updated_broadcaster = Some(futures_signals::signal::Broadcaster::new(receiver));
+        self.finish_build()
+    }
+
+    // defaults
+    fn default_sort(&self) -> Vec<String> {
+        vec!["by_recency".to_string(), "by_name".to_string()]
+    }
+
+    fn default_required_state(&self) -> Vec<(RoomEventType, String)> {
+        vec![
+            (RoomEventType::RoomEncryption, "".to_string()),
+            (RoomEventType::RoomTombstone, "".to_string()),
+        ]
+    }
+
+    /// Set the Syncing mode
+    pub fn sync_mode(&mut self, sync_mode: SlidingSyncMode) -> &mut Self {
+        self.sync_mode = Some(SyncMode::new(sync_mode));
+        self
+    }
+
+    /// Set the ranges to fetch
+    pub fn ranges<U: Into<UInt>>(&mut self, range: Vec<(U, U)>) -> &mut Self {
+        let mut new = self;
+        new.ranges =
+            Some(RangeState::new(range.into_iter().map(|(a, b)| (a.into(), b.into())).collect()));
+        new
+    }
+
+    /// Set the ranges to fetch
+    pub fn add_range<U: Into<UInt>>(&mut self, from: U, to: U) -> &mut Self {
+        let new = self;
+        let r = new.ranges.get_or_insert_with(|| RangeState::new(Vec::new()));
+        r.lock_mut().push((from.into(), to.into()));
+        new
+    }
+
+    /// Set the ranges to fetch
+    pub fn reset_ranges(&mut self) -> &mut Self {
+        let mut new = self;
+        new.ranges = None;
+        new
+    }
+    /// Set the limit of regular events to fetch for the timeline.
+    pub fn timeline_limit<U: Into<UInt>>(&mut self, timeline_limit: U) -> &mut Self {
+        let mut new = self;
+        new.timeline_limit = Some(Some(timeline_limit.into()));
+        new
+    }
+
+    /// Reset the limit of regular events to fetch for the timeline. It is left
+    /// to the server to decide how many to send back
+    pub fn no_timeline_limit(&mut self) -> &mut Self {
+        let mut new = self;
+        new.timeline_limit = None;
+        new
+    }
+}
+
+enum InnerSlidingSyncViewRequestGenerator {
+    FullSync(u32, u32), // current position, batch_size
```

Comment:
Why not naming the fields?
```suggestion
    FullSync { position: u32, batch_size: u32 }, // current position, batch_size
```